--- pcre2.h.in	2023-12-02 08:48:56
+++ pcre2.h	2024-02-15 14:11:12
@@ -41,10 +41,10 @@
 
 /* The current PCRE version information. */
 
-#define PCRE2_MAJOR           @PCRE2_MAJOR@
-#define PCRE2_MINOR           @PCRE2_MINOR@
-#define PCRE2_PRERELEASE      @PCRE2_PRERELEASE@
-#define PCRE2_DATE            @PCRE2_DATE@
+#define PCRE2_MAJOR           10
+#define PCRE2_MINOR           43
+#define PCRE2_PRERELEASE      -RC1
+#define PCRE2_DATE            2023-12-27
 
 /* When an application links to a PCRE DLL in Windows, the symbols that are
 imported have to be identified as such. When building PCRE2, the appropriate
